### DISCUSSION TITLE
Camera Orientation when Swimming

### DISCUSSION THREAD

**Discussion by sommerlost** [Sixze/ALS-Refactored#437] [Q&A]
I've been working on implementing Oceaography with ALS, I have everything working except for orienting player movement to the camera when the character enters the water.  I know this has to do with the locomotion mode, as other users have mentioned that you need to set it to Grounded or InAir. 

When I enter the swimming volume with ALS debug on, the movement mode is set to 'None'. After searching the source, I only see one call to `GetCharacterMovement()->SetMovementMode(MOVE_None);`, which occurs in: `void AAlsCharacter::StartRagdollingImplementation()`. So I'm confused as to how the movement mode is set to 'None' besides entering the ragdoll state?

**Comment by sommerlost** [comment]
After a little more tinkering this morning, I've come up with a solution. But I'm not sure if it's the best solution. On AlsCharacter.cpp

I added an entry for `MOVE_None` that set the LocomotionMode to Grounded, this allowing the character to rotate properly in the water:

```
void AAlsCharacter::OnMovementModeChanged(const EMovementMode PreviousMovementMode, const uint8 PreviousCustomMode)
{
	// Use the character movement mode to set the locomotion mode to the right value. This allows you to have a
	// custom set of movement modes but still use the functionality of the default character movement component.

	switch (GetCharacterMovement()->MovementMode)
	{
		case MOVE_Walking:
		case MOVE_NavWalking:
			SetLocomotionMode(AlsLocomotionModeTags::Grounded);
			break;

		case MOVE_Flying:
			SetLocomotionMode(AlsLocomotionModeTags::Grounded);
			break;

		case MOVE_Falling:
			SetLocomotionMode(AlsLocomotionModeTags::InAir);
			break;

		case MOVE_Swimming:
			//SetLocomotionMode(AlsLocomotionModeTags::InAir);
			SetLocomotionMode(AlsLocomotionModeTags::Grounded);
			break;

		case MOVE_None:
			SetLocomotionMode(AlsLocomotionModeTags::Grounded);
			break;

		default:
			SetLocomotionMode(FGameplayTag::EmptyTag);
			break;
	}

	Super::OnMovementModeChanged(PreviousMovementMode, PreviousCustomMode);
}
```
I'd prefer to have the volume set the locomotion mode to swim instead of none, as I'm sure this could cause some issues elsewhere, but I'm not sure how to progress on that. Does anyone have any pointers? Thanks!

**Reply by MostSinister** [reply]
I am running into the same problem with a different swim component and i had the same exact solution. But unfortunately setting the default flying mode to grounded fixes my orientation issue but causes other problems down the line. For instance i think on becoming grounded stops all movement for a sec which prevent me from sinking into the water like it should. 

Anyways im still investigating with my limited skills but im very interested in any other info anyone has on this.

thank you for sharing.
